src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala

```diff
@@ -281,9 +362,70 @@ private[pulsar] case class PulsarMetadataReader(
     }
   }
 
+  def actualOffsets(
+      offset: PerTopicOffset,
+      pollTimeoutMs: Int,
+      reportDataLoss: String => Unit): Map[String, MessageId] = {
+
+    offset match {
+      case so: SpecificPulsarOffset => fetchCurrentOffsets(so, pollTimeoutMs, reportDataLoss)
+      case st: SpecificPulsarStartingTime => fetchCurrentOffsets(st, pollTimeoutMs, reportDataLoss)
+      case t => throw new IllegalArgumentException(s"not supported offset type: $t")
+    }
+  }
+
+  def fetchCurrentOffsets(
+      time: SpecificPulsarStartingTime,
+      pollTimeoutMs: Int,
+      reportDataLoss: String => Unit): Map[String, MessageId] = {
+
+    time.topicTimes.map { case (tp, time) =>
+      val actualOffset =
+        if (time == PulsarProvider.EARLIEST_TIME) {
+          MessageId.earliest
+        } else if (time == PulsarProvider.LATEST_TIME) {
+          PulsarSourceUtils.seekableLatestMid(admin.topics().getLastMessageId(tp))
+        } else {
+          assert (time > 0, s"time less than 0: $time")
+          if (client == null) {
+            client = PulsarClient.builder().serviceUrl(serviceUrl).build()
+          }
+          val consumer = client
+            .newConsumer()
+            .topic(tp)
+            .subscriptionName(s"spark-pulsar-${UUID.randomUUID()}")
+            .subscriptionType(SubscriptionType.Exclusive)
+            .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+            .subscribe()
+
+          var earliestMessage: Message[Array[Byte]] = null
+          earliestMessage = consumer.receive(pollTimeoutMs, TimeUnit.MILLISECONDS)
+          if (earliestMessage == null) {
+            MessageId.earliest
+          } else {
+            val earliestId = earliestMessage.getMessageId
+
+            consumer.seek(time)
+            var msg: Message[Array[Byte]] = null
+            msg = consumer.receive(pollTimeoutMs, TimeUnit.MILLISECONDS)
```

Comment:
@sijie This `msg` may still have a publish time ealier than `time`.

Reply:
maybe we can add a setting in connector config to let connector skip the messages whose publish time is earlier than the given `time`.